File: app/src/main/java/com/meteorhead/leave/database/realm/LeaveRealmService.java
package com.meteorhead.leave.database.realm;

import com.meteorhead.leave.database.dbabstract.LeaveDbService;
import com.meteorhead.leave.database.dbabstract.base.DatabaseCallback;
import com.meteorhead.leave.database.dbabstract.base.DatabaseCallbackQuery;
import com.meteorhead.leave.database.realm.base.RealmService;
import com.meteorhead.leave.database.realm.base.interfaces.RealmCallback;
import com.meteorhead.leave.models.Leave;
import com.meteorhead.leave.models.LeaveFields;

import java.sql.Date;
import java.util.List;

import io.realm.Realm;
import io.realm.RealmChangeListener;
import io.realm.RealmResults;
import io.realm.Sort;
import rx.functions.Action0;
import rx.schedulers.Schedulers;

/**
 * Created by Lenovo on 2016-09-16.
 */

public class LeaveRealmService extends RealmService<Leave> implements LeaveDbService {

    public LeaveRealmService() {
        super(Leave.class, getMainThreadRealmInstance());
    }

    public LeaveRealmService(Realm realm) {
        super(Leave.class, realm);
    }

    public RealmResults<Leave> findLeaveBetweenDates(final Date startDate, final Date endDate) {
        return getServiceRealm().where(Leave.class)
                .between(LeaveFields.DATE_START, startDate, endDate)
                .between(LeaveFields.DATE_END, startDate, endDate)
                .findAll();
    }

    @Override
    public RealmResults<Leave> getAllLeaves() {
        return getServiceRealm().where(Leave.class)
                .findAllSorted(LeaveFields.DATE_START, Sort.ASCENDING);
    }

    @Override
    public RealmResults<Leave> getAllLeavesFuture() {
        return getServiceRealm().where(Leave.class)
                .findAllSortedAsync(LeaveFields.DATE_START, Sort.ASCENDING);
    }

    @Override
    public Leave getLeaveById(int id) {
        return getServiceRealm().where(Leave.class).equalTo(LeaveFields.ID, id).findFirst();
    }

    @Override
    public void addOrUpdate(final Leave leaveToAdd, DatabaseCallback callback) throws IllegalArgumentException{
        RealmCallback realmCallback = convertCallback(callback);

        int id = leaveToAdd.getId();
        if(leaveToAdd.getId() == -1) {
            id = (int) getNextPrimaryKey(getServiceRealm());
        }
        final int primaryKey = id;
        getServiceRealm().executeTransactionAsync(realm -> {
            leaveToAdd.setId(primaryKey);
            realm.insertOrUpdate(leaveToAdd);
        }, realmCallback, realmCallback);
    }

    @Override
    public void insertLeave(final Leave leaveToInsert, DatabaseCallback callback) {
        RealmCallback realmCallback = convertCallback(callback);
        getServiceRealm().executeTransactionAsync(realm -> realm.copyToRealm(leaveToInsert), realmCallback, realmCallback);
    }

    @Override
    public void insertLeaves(final List<Leave> leavesToInsert) {
        getServiceRealm().executeTransaction(realm -> realm.insert(leavesToInsert));
    }

    @Override
    public void removeLeave(final Leave leaveToRemove) {
        getServiceRealm().executeTransaction(realm -> leaveToRemove.deleteFromRealm());
    }

    @Override
    public void removeLeaves(final List<Leave> leavesToRemove) {
        getServiceRealm().executeTransaction(realm -> {
            for (Leave leave : leavesToRemove) {
                leave.deleteFromRealm();
            }
        });
    }

    @Override
    public void finish() {
        close();
    }

    @Override
    public Leave copy(Leave leave) {
        return getServiceRealm().copyFromRealm(leave);
    }

    @Override
    public List<Leave> copy(List<Leave> leaves) {
        return getServiceRealm().copyFromRealm(leaves);
    }


}

File: app/src/main/java/com/meteorhead/leave/database/realm/LeaveRealmService.kt
package com.meteorhead.leave.database.realm

import com.meteorhead.leave.database.dbabstract.LeaveDbService
import com.meteorhead.leave.database.dbabstract.base.DatabaseCallback
import com.meteorhead.leave.database.realm.base.RealmService
import com.meteorhead.leave.database.realm.base.interfaces.RealmCallback
import com.meteorhead.leave.data.Leave
import io.realm.Realm
import io.realm.RealmResults
import io.realm.Sort
import java.sql.Date

/**
 * Created by Lenovo on 2016-09-16.
 */

class LeaveRealmService : RealmService<Leave>, LeaveDbService {

    constructor() : super(Leave::class.java, null)

    constructor(realm: Realm) : super(Leave::class.java, realm)

    override fun findLeaveBetweenDates(startDate: Date, endDate: Date): RealmResults<Leave> {
        return serviceRealm.where(Leave::class.java)
            .between(Leave::dateStart.name, startDate, endDate)
            .between(Leave::dateEnd.name, startDate, endDate)
            .findAll()
    }

    override fun getAllLeaves(): RealmResults<Leave> {
        return serviceRealm.where(Leave::class.java)
            .findAllSorted(Leave::dateStart.name, Sort.ASCENDING)
    }

    override fun getAllLeavesFuture(): RealmResults<Leave> {
        return serviceRealm.where(Leave::class.java)
            .findAllSortedAsync(Leave::dateStart.name, Sort.ASCENDING)
    }

    override fun getLeaveById(id: Int): Leave {
        return serviceRealm.where(Leave::class.java).equalTo(Leave::id.name, id).findFirst()
    }

    @Throws(IllegalArgumentException::class)
    override fun addOrUpdate(leaveToAdd: Leave, callback: DatabaseCallback) {
        val realmCallback = convertCallback(callback)

        var id = leaveToAdd.id
        if (leaveToAdd.id == -1) {
            id = getNextPrimaryKey(serviceRealm) as Int
        }
        val primaryKey = id
        serviceRealm.executeTransactionAsync(Realm.Transaction { realm ->
            leaveToAdd.id = primaryKey
            realm.insertOrUpdate(leaveToAdd)
        }, realmCallback, realmCallback)
    }

    override fun insertLeave(leaveToInsert: Leave, callback: DatabaseCallback) {
        val realmCallback = convertCallback(callback)
        serviceRealm.executeTransactionAsync(
            Realm.Transaction { realm ->
                realm.copyToRealm(leaveToInsert)
            }, realmCallback, realmCallback)
    }

    override fun insertLeaves(leavesToInsert: List<Leave>) {
        serviceRealm.executeTransaction { realm -> realm.insert(leavesToInsert) }
    }

    override fun removeLeave(leaveToRemove: Leave) {
        serviceRealm.executeTransaction { realm -> leaveToRemove.deleteFromRealm() }
    }

    override fun removeLeaves(leavesToRemove: List<Leave>) {
        serviceRealm.executeTransaction { realm ->
            for (leave in leavesToRemove) {
                leave.deleteFromRealm()
            }
        }
    }

    override fun finish() {
        close()
    }

    override fun copy(leave: Leave): Leave {
        return serviceRealm.copyFromRealm(leave)
    }

    override fun copy(leaves: List<Leave>): List<Leave> {
        return serviceRealm.copyFromRealm(leaves)
    }

}
